app/Http/Controllers/PostsController.php

```diff
@@ -120,15 +120,18 @@ public function store(PostRequest $request)
      */
     public function show(Post $post)
     {
-        // Only allow an admin or the user who owns the post to see their own unapproved posts.
-        if ($post->status != 'accepted') {
-            if (is_staff_user() || auth()->id() === $post->northstar_id) {
+        // If there is no user (anonymous request), only return a post if it is approved.
+        if (auth()->id() === null) {
```

Comment:
Since this is an anonymous request, there is no `$user` being sent to the policy and it seems to break and never reach the `show` policy logic. 

The only way I could get this to work (kind of - still having trouble having the `AuthorizationException` being thrown, getting a `Undefined variable: errors (View: /home/vagrant/DoSomething.org/rogue/resources/views/layouts/master.blade.php) (View: /home/vagrant/DoSomething.org/rogue/resources/views/layouts/master.blade.php)` error when the post's `status` is `rejected` or `pending`) was to do something like this but this seems confusing to add this logic in addition to the policy. 

Any ideas as to how to send an anonymous user to the policy?

Reply:
do we want to do something like [this](https://laravel.io/forum/07-28-2014-force-special-user-instance-for-anonymous-user)?

Reply:
Note: looks like this is broken with the current code as well (throws the same error for an anonymous request for a `pending` or `rejected` post).

Reply:
@chloealee could we do something like this within the policy? https://laracasts.com/discuss/channels/laravel/user-laravel-policies-with-anonymous-users?page=1

Reply:
Like using that `before()` method to decide what to do with the anon user?

Reply:
Hmm, that's frustrating! I'm surprised there isn't better support for that in the framework.

> I tried that but it seems like it doesn't even reach the `before()` method if there is no `$user` :-/

Yeah, it looks like Laravel [short-circuits and return `false` here](https://github.com/laravel/framework/blob/39165ddea22cf837a34a8b01e4974cebb5c1af6d/src/Illuminate/Auth/Access/Gate.php#L299-L301) if there's no user. I wonder if the cleanest solution would be to override the `Gate` class (by binding a custom implementation in our AppServiceProvider) and override `raw` to skip that conditional?

Reply:
@DFurnes I think that's a great idea! That's what I was thinking too after reading [this post](https://stackoverflow.com/questions/35317909/authorization-and-policies-for-guest-users) on Stackoverflow - I think they are suggesting something similar?

Reply:
@DFurnes that sounds good to me! How would we override the `Gate` class in the `AppServiceProvider`? I tried to add a `raw` function in [that file](https://github.com/DoSomething/rogue/blob/master/app/Providers/AppServiceProvider.php) but I don't think that is the correct way to do it?

Reply:
We'd be making our own version of [the `Gate` contract](https://github.com/laravel/framework/blob/5.5/src/Illuminate/Contracts/Auth/Access/Gate.php) (that extends the [default one](https://github.com/laravel/framework/blob/5.5/src/Illuminate/Auth/Access/Gate.php), but overrides just that `raw` method). You can then [bind our custom implementation](https://laravel.com/docs/5.6/container#binding) in the app service provider.

Reply:
Thanks @DFurnes ! Sorry I'm having a hard time wrapping my head around this!  So do we just copy the [`Gate` contract](https://github.com/laravel/framework/blob/5.5/src/Illuminate/Contracts/Auth/Access/Gate.php) and add a `raw` method there? Where should this file go? I don't think we want it to be in `vendor/laravel/framework/src/Illuminate/Contracts/Auth/Access` since it is custom?

Reply:
I _think_ you can put it in any sensible place. maybe a `App/Contract/Auth/` directory. And then in the binding that dave pointed to, you specify which file to point to.

Reply:
Yep, I think `app/Auth/Gate.php` is a fine place for this!

Reply:
> So do we just copy the [`Gate` contract](https://github.com/laravel/framework/blob/5.5/src/Illuminate/Contracts/Auth/Access/Gate.php) and add a `raw` method there? Where should this file go? I don't think we want it to be in `vendor/laravel/framework/src/Illuminate/Contracts/Auth/Access` since it is custom?

We'll be making a new implementation of that same contract! The contract describes what the class can do, and it's what allows Laravel to let us provide our own implementation (since it knows it must have the same methods, then).

So we want to make a new `Gate` implementation (let's say in `app/Auth/Gate.php`) that implements that contract, extends the default (so we get all the existing behavior), and then just overrides that one method to work a little differently. So like this:

```php
<?php

namespace Rogue\Auth;

use Illuminate\Auth\Access\Gate;
use Illuminate\Contracts\Auth\Access\Gate as GateContract;


class Gate extends Gate implements GateContract
{
    protected function raw($ability, $arguments = [])
    {
        // copy-paste the implementation from Laravel's Gate, but without that
        // first conditional that we don't want!
    }
}

```

Reply:
this is super helpful, thanks @DFurnes and @sbsmith86 ! When binding, I think I would do something like this: 
```
$this->app->bind(
            'Illuminate\Contracts\Auth\Access\Gate',
            'Rogue\Auth\Gate'
        );
```
I'm getting an error though - is this not the way to bind? I'll push up what I've got now!